pruner/state.go

```diff
@@ -0,0 +1,81 @@
+package pruner
+
+import (
+	"fmt"
+	"sync"
+
+	quamina "quamina/lib"
+)
+
+// State represents the required capabilities for maintaining the set
+// of live patterns.
+type State interface {
+	Add(x quamina.X, pattern string) error
+	Del(x quamina.X) (bool, error)
+
+	// Iterate calls the given function for every stored pattern.
+	Iterate(func(x quamina.X, pattern string) error) error
+
+	// Get returns the pattern for the given X.
+	//
+	// Since a pattern can't be the empty string, that zero value
+	// indicates no corresponding pattern.
+	Get(x quamina.X) (string, error)
+}
+
+// MemState is a State that is just a map (with a RWMutex).
+type MemState struct {
```

Comment:
RWMutex seems to have a bit of a bad rep. Maybe a few words of comment about why this, rather than for example figuring out a way to do an atomic update like smallTable does?

Reply:
Yah, some quick commentary does seem helpful here. Since the application can provide its own `State` implementation, I was leaning in favor of simplicity here. [`sync.Map`](https://pkg.go.dev/sync#Map) seems more distasteful than `RWMutex`. I wanted a `map`-based implementation for simplicity -- or something similar that's built from the standard library. But I failed to come up with any better on balance. In particular, I don't see how to pull off some sort of atomic update of a field while keeping things simple.

FWIW, I recall some interesting discussion a few years back about what assignments were and were not considered atomic -- and on what platforms, etc. Official docs [here](https://go.dev/ref/mem), I guess. (Bringing out [`sync.atomic`](https://pkg.go.dev/sync/atomic) didn't obviously seem to help me anyway.)

Maybe some cleverness for this stock implementation can evolve.